src/browser/file-recovery-service.js

```diff
@@ -0,0 +1,65 @@
+'use babel'
+
+import {ipcMain} from 'electron'
+import crypto from 'crypto'
+import Path from 'path'
+import fs from 'fs-plus'
+
+export default class FileRecoveryService {
+  constructor (recoveryDirectory) {
+    this.recoveryDirectory = recoveryDirectory
+    this.recoveryPathsByWindowAndFilePath = new WeakMap()
+    this.crashListeners = new WeakSet()
+  }
+
+  start () {
+    ipcMain.on('will-save-path', this.willSavePath.bind(this))
+    ipcMain.on('did-save-path', this.didSavePath.bind(this))
+  }
+
+  willSavePath (event, path) {
+    if (!fs.existsSync(path)) {
+      // Unexisting files won't be truncated/overwritten, and so there's no data to be lost.
+      return
+    }
+
+    const window = event.sender
+    const recoveryFileName = crypto.randomBytes(5).toString('hex')
+    const recoveryPath = Path.join(this.recoveryDirectory, recoveryFileName)
+    fs.writeFileSync(recoveryPath, fs.readFileSync(path))
```

Comment:
Okay, that sounds reasonable to me. Would it be worth using the original file basename as part of the filename so that they can be more easily identified?